exec/rpc/src/main/java/org/apache/drill/exec/rpc/AbstractRemoteConnection.java

```diff
@@ -224,4 +240,104 @@ public void close() {
     }
   }
 
+  /**
+   * Helps to add all the required security handler's after negotiation for encryption is completed.
+   * <p>Handler's before encryption is negotiated are:</p>
+   * <ul>
+   *    <li>  PROTOCOL_DECODER {@link ProtobufLengthDecoder} </li>
+   *    <li>  MESSAGE_DECODER {@link RpcDecoder}  </li>
+   *    <li>  PROTOCOL_ENCODER {@link RpcEncoder} </li>
+   *    <li>  HANDSHAKE_HANDLER {@link org.apache.drill.exec.rpc.BasicClient.ClientHandshakeHandler} OR
+   *                            {@link org.apache.drill.exec.rpc.BasicServer.ServerHandshakeHandler}  </li>
+   *    <li>  optional - IDLE_STATE_HANDLER {@link org.apache.drill.exec.rpc.BasicClient.IdlePingHandler} OR
+   *                   - TIMEOUT_HANDLER {@link org.apache.drill.exec.rpc.BasicServer.LoggingReadTimeoutHandler}  </li>
+   *    <li>  MESSAGE_HANDLER {@link org.apache.drill.exec.rpc.RpcBus.InboundHandler} </li>
+   *    <li>  EXCEPTION_HANDLER {@link RpcExceptionHandler} </li>
+   * </ul>
+   * <p>Handler's after encryption is negotiated are:</p>
+   * <ul>
+   *    <li>  LENGTH_DECODER_HANDLER {@link LengthFieldBasedFrameDecoder}
+   *    <li>  SASL_DECRYPTION_HANDLER {@link SaslDecryptionHandler}
+   *    <li>  PROTOCOL_DECODER {@link ProtobufLengthDecoder}
+   *    <li>  MESSAGE_DECODER {@link RpcDecoder}
+   *    <li>  SASL_ENCRYPTION_HANDLER {@link SaslEncryptionHandler}
+   *    <li>  CHUNK_CREATION_HANDLER {@link ChunkCreationHandler}
+   *    <li>  PROTOCOL_ENCODER {@link RpcEncoder}
+   *    <li>  HANDSHAKE_HANDLER {@link org.apache.drill.exec.rpc.BasicClient.ClientHandshakeHandler} OR
+   *                            {@link org.apache.drill.exec.rpc.BasicServer.ServerHandshakeHandler}
+   *    <li>  optional - IDLE_STATE_HANDLER {@link org.apache.drill.exec.rpc.BasicClient.IdlePingHandler} OR
+   *                   - TIMEOUT_HANDLER {@link org.apache.drill.exec.rpc.BasicServer.LoggingReadTimeoutHandler}
+   *    <li>  MESSAGE_HANDLER {@link org.apache.drill.exec.rpc.RpcBus.InboundHandler}
+   *    <li>  EXCEPTION_HANDLER {@link RpcExceptionHandler}
+   * </ul>
+   * <p>
+   *  If encryption is enabled ChunkCreationHandler is always added to divide the Rpc message into chunks of
```

Comment:
This is required since Encoded Message size cannot be bigger than 16MB, this handler will take care of that. Keeping a separate handler for chunking purpose also helps to keep the encryption and chunking logic separate. As far as performance degradation is concerned I don't think there should be any notable difference.